### PR TITLE
[19.07] Revert "dnsmasq: skip options that are not compiled in"

### PR DESCRIPTION
**Pull request by hansmi**
The revert contained in this pull request is also part of #1812 which hasn't been reviewed as of this writing. DNSSEC validation in OpenWRT 19.07 is _not working_ without this revert. The other change in #1812 would be nice to have, but isn't required.
<hr>

This reverts commit 04b45d3a31fac45c472ad3c31d98268d1c309763 and
0299a4b73e729504dfdc5c4563535db082bdf941. The changes caused dnsmasq to
silently ignore configured options instead of failing or at least
logging warnings.

In addition DNSSEC support was rendered unusable: The trust anchors are
loaded from a configuration file ("--conf-file=$TRUSTANCHORSFILE"), an
option which wasn't considered by the filter and thus caused dnsmasq
without DNSSEC support to not start (the trust anchor file is only
included with the "dnsmasq-full" package). In addition the plain
"--dnssec" option was not recognized by the filter.

The reverted change is better implemented as feature-specific code which
also issues warnings to the user if a feature is enabled by
configuration, but not available in the installed build.

### CODE REVIEW

**Comment by hauke** [comment]
@yousong  can you please comment on this.

**Comment by yousong** [comment]
Warn when skipping is good to have

However, I think the detection logic is required to workaround [FS#2281](https://bugs.openwrt.org/index.php?do=details&task_id=2281).  The issue reported there is a valid one.  We allow custom build of dnsmasq, and that could be rendered not usable when the feature detection and option skipping logic were stripped out.

> In addition DNSSEC support was rendered unusable: The trust anchors are
loaded from a configuration file ("--conf-file=$TRUSTANCHORSFILE"), an
option which wasn't considered by the filter and thus caused dnsmasq
without DNSSEC support to not start (the trust anchor file is only
included with the "dnsmasq-full" package).

This is expected as said in one of two notes in the to-be-reverted commit 

> Quote https://github.com/openwrt/openwrt/commit/0299a4b73e729504dfdc5c4563535db082bdf941
> - This provides a way out but users' cooperation is still needed.  E.g.
   option dnssec needs to be turned off, otherwise the service script
   will try to add --conf-file pointing to dnssec specific anchor file
   which dnsmasq lacking dnssec support will reject

As for

> In addition the plain
"--dnssec" option was not recognized by the filter.

It's a uci option, and you can turn it off.  Then no dnssec anchor config file will be loaded.

https://github.com/openwrt/openwrt/blob/17e2246eca901fec79fbac5b7a90f68f60785c4e/package/network/services/dnsmasq/files/dnsmasq.init#L959-L971

Thanks for your attention.   And sorry for the delay,  feel free to directly drop me a mail if I missed github notification for too long.

**Comment by hansmi** [comment]
> Warn when skipping is good to have
> 
> However, I think the detection logic is required to workaround [FS#2281](https://bugs.openwrt.org/index.php?do=details&task_id=2281). The issue reported there is a valid one. We allow custom build of dnsmasq, and that could be rendered not usable when the feature detection and option skipping logic were stripped out.

The commits rolled back by this PR tried to implement that, but did so in a very fragile and impactful way (see below). #1812 first proposed silently disabling DNSSEC support when not compiled in and, after a discussion, I changed it to warn at least. That mechanism can be extended to cover more features.

> > In addition DNSSEC support was rendered unusable: The trust anchors are
> > loaded from a configuration file ("--conf-file=$TRUSTANCHORSFILE"), an
> > option which wasn't considered by the filter and thus caused dnsmasq
> > without DNSSEC support to not start (the trust anchor file is only
> > included with the "dnsmasq-full" package).
> 
> This is expected as said in one of two notes in the to-be-reverted commit

Breaking DNSSEC support for `dnssec-full` as distributed for the upstream `x86_64` build is expected?

> > Quote [0299a4b](https://github.com/openwrt/openwrt/commit/0299a4b73e729504dfdc5c4563535db082bdf941)
> > 
> > * This provides a way out but users' cooperation is still needed.  E.g.
> >   option dnssec needs to be turned off, otherwise the service script
> >   will try to add --conf-file pointing to dnssec specific anchor file
> >   which dnsmasq lacking dnssec support will reject
> 
> As for
> 
> > In addition the plain
> > "--dnssec" option was not recognized by the filter.
> 
> It's a uci option, and you can turn it off. Then no dnssec anchor config file will be loaded.

Yes, but please read your code again. It checks for `dnssec-*`, _not_ `dnssec | dnssec-*` or something comparable, and it doesn't consider the trust-anchor file either.

**Comment by yousong** [comment]
> Yes, but please read your code again. It checks for `dnssec-*`, _not_ `dnssec | dnssec-*` or something comparable, and it doesn't consider the trust-anchor file either.

`--dnssec` is not checked in the initscript because dnsmasq accepts it (does not quit) regardless of whether DNSSEC support is there or not.

If "dnssec" uci option is set to off, initscript will not add `--conf-file=$TRUSTANCHORSFILE`.  Yes, arguments included such way are not checked by current implementation, but at the moment I do not see it to cause problem because

 - If dnsmasq has dnssec support, it will accept the `trust-anchor=` option in the included config file just fine.
 - If dnsmasq lacks dnssec support, you turn that uci option off and dnsmasq will not be bothered at all.

And the current code does take into account `trust-anchor` option, though the initscript does not directly generate such option at the moment.

I still do not see how it can break  dnsmasq-full build with dnssec support.

I'd like note that current implementation of guarding against unsupported option only at `xappend` has be benefit of being at least less intrusive in my opinion.  Just think about the fact the dnsmasq initscript is now a 1172 lines of shell script with 64 direct call to the `xappend` function.  This matters

**Comment by hansmi** [comment]
> --dnssec is not checked in the initscript because dnsmasq accepts it (does not quit) regardless of whether DNSSEC support is there or not.

That's incorrect:

```
# dnsmasq --version
Dnsmasq version 2.80  Copyright (c) 2000-2018 Simon Kelley
Compile time options: IPv6 GNU-getopt no-DBus no-i18n no-IDN DHCP no-DHCPv6 no-Lua TFTP no-conntrack no-ipset no-auth no-DNSSEC no-ID loop-detect inotify dumpfile

# dnsmasq --dnssec; echo $?
dnsmasq: DNSSEC not available: set HAVE_DNSSEC in src/config.h
1
```

I also tested supplying `dnssec` via the config file. The behaviour is the same.

> If dnsmasq lacks dnssec support, you turn that uci option off and dnsmasq will not be bothered at all.

The point of #1812 is to not make that necessary, but obviously not of this PR. Currently the behaviour in 19.07.0-rc2 is as follows when enabling DNSSEC with the lighter `dnsmasq` package (not `dnsmasq-full`):

```
dnsmasq: DNS rebinding protection is active, will discard upstream RFC1918 responses!
dnsmasq: Allowing 127.0.0.0/8 responses
dnsmasq[2948]: cannot read /usr/share/dnsmasq/trust-anchors.conf: No such file or directory
dnsmasq[2948]: FAILED to start up
```

Before your changes the error message was "DNSSEC not available" (as above) which is definitely more user-friendly.

> I still do not see how it can break dnsmasq-full build with dnssec support.

You're right, I must've misremembered that part. This code had been waiting for a review since November 10, 2019 (it's part of #1812, I only cherry-picked the revert commit for this PR).

I used [`openwrt-19.07.0-rc2-x86-64-combined-ext4.img.gz`](https://downloads.openwrt.org/releases/19.07.0-rc2/targets/x86/64/openwrt-19.07.0-rc2-x86-64-combined-ext4.img.gz) as a start with the following commands:

```
opkg update

# As needed
opkg remove dnsmasq
opkg install dnsmasq-full

# Configure DNSSEC (0/1)
uci set 'dhcp.@dnsmasq[0].dnssec=1'
uci commit dhcp

# Restart dnsmasq
/etc/init.d/dnsmasq restart

# Check log
logread | tail -n 100
```

**Comment by yousong** [comment]
> > --dnssec is not checked in the initscript because dnsmasq accepts it (does not quit) regardless of whether DNSSEC support is there or not.
> 
> That's incorrect:
> 
> ```
> # dnsmasq --version
> Dnsmasq version 2.80  Copyright (c) 2000-2018 Simon Kelley
> Compile time options: IPv6 GNU-getopt no-DBus no-i18n no-IDN DHCP no-DHCPv6 no-Lua TFTP no-conntrack no-ipset no-auth no-DNSSEC no-ID loop-detect inotify dumpfile
> 
> # dnsmasq --dnssec; echo $?
> dnsmasq: DNSSEC not available: set HAVE_DNSSEC in src/config.h
> 1
> ```
> 
> I also tested supplying `dnssec` via the config file. The behaviour is the same.
> 
> >

My statement of "does not quit" is wrong.  I intended to mean that dnsmasq recognizes "--dnssec" option regardless of whether dnssec support is builtin.  "--dnssec is for DNSSEC but we do not have DNSSEC support", vs. "unknown argument --dnssec"

**Comment by ynezz** [comment]
tl;dr, whats the status of this @yousong ? :)

**Comment by yousong** [comment]
In the least I think we agreed that the commit to be reverted in this pull request does NOT "break dnsmasq-full build with dnssec support."  It's still my opinion that the revert is not necessary.

Thank you for your time and efforts.
